c/restService.c

```diff
@@ -0,0 +1,266 @@
+  jsonEndObject(out);
+  jsonStartObject(out, NULL);
+  jsonAddString(out, "href", "/server/agent/environment");
+  jsonAddString(out, "rel", "environment");
+  jsonAddString(out, "type", "GET");
+  jsonEndObject(out);
+  jsonEndArray(out);
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+void respondWithLogLevels(HttpResponse *response){
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  JsonObject *logLevels = jsonObjectGetObject(serverConfig, "logLevels");
+  setResponseStatus(response, 200, "OK");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  if(logLevels){
+    jsonPrintObject(out, logLevels);
+  }
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+void respondWithServerEnvironment(HttpResponse *response){
+  /*Information about parameters for smc_func: https://www.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.erbb700/smfp.htm#smfp*/
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  struct utsname unameRet;
+  uname(&unameRet);
+  char pid[64];
+  char ppid[64];
+  char dp[64];
+  char cpu_u[64];
+  char mvs_u[64];
+  char ziip_u[64];
+  char zaap_u[64];
+  char* buffer;
+  int rc = 0;
+  int reqtype = 0x00000005; //fullword. request type
+  int rectype = 0x0000004F; //SMF record type, only type 79 is supported
+  int subtype = 0x00000009; //SMF record subtype
+  int bufferlen = 247488; /* length of SMF record buffer */
+  int cpu_util = 0x00000000;
+  int demand_paging = 0x00000000;
+  int options = 0x00000000;
+  int mvs_srm = 0x00000000;
+  int zaap_util = 0x00000000;
+  int ziip_util = 0x00000000;
+  buffer = (char *)safeMalloc(bufferlen, "buffer");
+  memset(buffer, 0, bufferlen);
+  smf_func = (EXSMFI *)fetch("ERBSMFI");
+  rc = (*smf_func)(&reqtype,
+               &rectype,
+               &subtype,
+               buffer,
+               &bufferlen,
+               &cpu_util,
+               &demand_paging,
+               &options,
+               &mvs_srm,
+               &zaap_util,
+               &ziip_util);
+  sprintf(pid, "%d", getpid());
+  sprintf(ppid, "%d", getppid());
+  sprintf(dp, "%d", demand_paging);
+  sprintf(cpu_u, "%d%", cpu_util);
+  sprintf(mvs_u, "%d%", mvs_srm);
+  sprintf(zaap_u, "%d%", zaap_util);
+  sprintf(ziip_u, "%d%", ziip_util);
+  setResponseStatus(response, 200, "OK");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  jsonAddString(out, "logDirectory", getenv("ZSS_LOG_FILE"));
+  jsonAddString(out, "agentName", "zss");
+  jsonAddString(out, "agentVersion", productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *env_var = strdup(*environ);
+  int i = 1;
+  for (; env_var; i++) {
+    char *var_name = strtok(env_var, "=");
```

Comment:
What if there is no ```"="``` in the string?

Reply:
I do not believe this is a possible scenario for environment variables.  There must be at a minimum one "=" unless I am mistaken